File: controllers/langtags/language_controller.py
from repositories.langtags.language_repository import LanguageRepository
from models.langtags.language import Language

class LanguageController:
    def __init__(self, repository:LanguageRepository, model:Language):
        self.repository = repository
        self.model = model

    # Funciones genericas
    def get_columns(self):
        return self.repository.table.get_columns()

    def get_column_values(self):
        return self.repository.table.get_column_values()

    # Especificas
    def get_row(self, code_id) -> bool:
        row = self.repository.get_row( code_id )
        if len(row) > 0:
            self.model.language_id = row[0]
            self.model.code = row[1]
            self.model.created_at = row[2]
            self.model.updated_at = row[3]
            self.model.deleted_at = row[4]
            return True
        else:
            return False

    def get_code_row(self, code:str) -> bool:
        code_id = self.repository.get_code_id( code )
        return self.get_row( code_id )

    def update(self, language_id, code) -> bool:
        if self.repository.update_code( language_id, code ):
            self.get_row( language_id )
            return True
        return False

    def insert(self, code) -> bool:
        if isinstance(code, str):
            if self.repository.insert_code( code ):
                self.get_code_row( code )
                return True
        return False

    def save(self, language_id, code):
        if self.repository.exists( language_id ):
            return self.update( language_id, code)
        return self.insert( code )

    def is_deleted(self, language_id):
        return self.repository.is_deleted( language_id )

    def is_model_deleted(self):
        return self.repository.is_deleted( self.model.language_id )

File: Controllers/LangTags/LanguageController.cs
using System.Collections.Generic;

// Logger
using Microsoft.Extensions.Logging;

// Essentials
using Repositories.LangTags;
using Entities.LangTags;

namespace Controllers.LangTags {
    public class LanguageController : ITableController
    {
        // Variables
        private LanguageRepository _repository;
        private ILogger _logger;
        private LanguageEntity _entity;

        // Constructor
        public LanguageController(
            LanguageRepository repository, LanguageEntity entity, ILogger logger)
        {
            _repository = repository;
            _entity = entity;
            _logger = logger;
        }

        // Methods
        public void Create(string code){
            try {
                bool exists = _repository.ExistsByCode(code);
                if (!exists){
                    _repository.Insert(code, true);
                    _logger.LogInformation($"Code inserted `{code}`");
                }
            } catch (Exception e){
                _logger.LogError(e, $"Error inserted `{code}`");
            }
        }

        public void Update(int id, string code){
            try {
                bool exists = _repository.ExistsById(id);
                if (!exists){
                    _repository.Update(id, code, true);
                    _logger.LogInformation($"Code updated `{code}`");
                }
            } catch (Exception e){
                _logger.LogError(e, $"Error updated `{code}`");
            }
        }

        public void Save(string code, bool isActive=true, int? languageId=null) {
            try {
                _repository.Save(
                    code:code, isActive:isActive, languageId:languageId
                );
            } catch (Exception e){
                _logger.LogError(e, $"Error saving: `{code}`");
            }
        }

        public void Delete(int id){
            try {
                bool exists = _repository.ExistsById(id);
                if (!exists){
                    _repository.Deactivate(id);
                    _logger.LogInformation($"Code deleted `{id}`");
                }
            } catch (Exception e){
                _logger.LogError(e, $"Error deleted `{id}`");
            }
        }

        public void Activate(int id){
            try {
                bool exists = _repository.ExistsById(id);
                if (!exists){
                    _repository.Activate(id);
                    _logger.LogInformation($"Code activated `{id}`");
                }
            } catch (Exception e){
                _logger.LogError(e, $"Error activated `{id}`");
            }
        }
        public string GetCodeById(int id){
            try {
                return _repository.GetCodeById(id);
            } catch {
                return "";
            }
        }
        public int GetIdByCode(string code){
            try {
                return _repository.GetIdByCode(code);
            } catch {
                return -1;
            }
        }

        public bool ExistsById(int id){
            return _repository.ExistsById(id);
        }

        public bool ExistsByCode(string code){
            return _repository.ExistsByCode(code);
        }

        public string[] GetColumnNames() {
            try {
                return _repository.Table.GetColumnNames();
            } catch
            {
                return new string[0];
            }
        }

        public List<string[]> GetRowValues(){
            try {
                return _repository.Table.GetRowValues();
            } catch
            {
                return new List<string[]>();
            }
        }

    }
}
